app/services/charges/charge_models/prorated_graduated_service.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+module Charges
+  module ChargeModels
+    class ProratedGraduatedService < Charges::ChargeModels::BaseService
+      protected
+
+      def ranges
+        properties['graduated_ranges']&.map(&:with_indifferent_access)
+      end
+
+      def compute_amount
+        full_units = per_event_aggregation_result.event_aggregation
+        prorated_units = per_event_aggregation_result.event_prorated_aggregation
+
+        index = 0
+        full_sum = 0
+        overflow = 0
+        ranges.reduce(0) do |result_amount, range|
```

Comment:
Is there a way to break this logic into smaller methods? it's really hard to read :/

Reply:
Refactored a bit, hope so that it's more readable now:)